packages/txservice/src/dispatch.ts

```diff
@@ -0,0 +1,195 @@
+import { BigNumber } from "ethers";
+import { BaseLogger } from "pino";
+import PriorityQueue from "p-queue";
+
+import { ChainRpcProvider } from "./provider";
+import { GasPrice, WriteTransaction } from "./types";
+import { Transaction } from "./transaction";
+import { TimeoutError, TransactionReverted } from "./error";
+
+// TODO: sep file
+// Thread safe stack manager.
+class TransactionBuffer {
+  // Both of the following are functionally treated as stacks (LIFO).
+  private validated: Transaction[] = [];
+  private pending: Transaction[] = [];
+  // Should be used to access either of the stacks.
+  private readonly accessQueue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
```

Comment:
I don't understand why you need to serialize access so the transactions, or really what the `validated` stack is for?

Reply:
A: serializing access here was the equivelent of having a "thread-safe" flag for accessing something, which is a common practice with multithreading in other languages
It serializes all reads, because time is sort of of the essence here

That being said, it may be *totally* unnecessary. I think worth trying with and without for sure

B: validated is to provide lookback to get the current nonce. to get the current nonce, we look back through pending and then validated. If nothing, we just call provider.getTransactionCount